File: Source/Windows/Remote Control Server/Network.vb
Imports System.Net
Imports System.Threading
Imports System.IO

Public Module Network

    Public Const localHost As String = "127.0.0.1"
    Public hostIps As List(Of String)
    Public anyIpEndPoint As New System.Net.IPEndPoint(System.Net.IPAddress.Any, 0)

    Public tcp As TCP
    Public udp As UDP

    Public commandCount As Long = 0

    Public Sub initialize()
        Logger.add("Initializing network")

        hostIps = New List(Of String)
        getHostIps()

        tcp = New TCP
        tcp.initialize()

        udp = New UDP
        udp.initialize()

        Dim listenThread As Thread = New Thread(New ThreadStart(AddressOf checkListenersRunning))
        listenThread.Start()
    End Sub

    Public Function checkListenersRunning() As Boolean
        If tcp.isListening And udp.isListening Then
            Return True
        Else
            Logger.add("Network initialization failed, not all listeners could be started.")
            Server.gui.showErrorDialog("Error", "Network initialization failed, not all listeners could be started." _
                                  & vbNewLine & "Make sure that only one instance of the Remote Control Server is running.")
            Return False
        End If
    End Function

    Public Sub getHostIps()
        Logger.add("Active Wi-Fi network adapters:")

        For Each ipAddress As IPAddress In Dns.GetHostByName(Dns.GetHostName()).AddressList
            Dim ip As String = ipAddress.ToString
            hostIps.Add(ip)
            Logger.add(" - " & ip)
        Next
    End Sub

    Public Function getServerIp() As String
        Dim ip As String = "Unknown"
        Try
            ip = hostIps(hostIps.Count - 1)

            If hostIps.Contains(Remote.lastCommand.destination) Then
                ip = lastCommand.destination
            End If
        Catch ex As Exception

        End Try
        Return ip
    End Function

    Private Sub sendCommandThread(ByVal command As Command)
        Select Case command.priority
            Case RemoteControlServer.Command.PRIORITY_LOW
                udp.sendData(command)
            Case RemoteControlServer.Command.PRIORITY_MEDIUM
                tcp.sendData(command)
            Case RemoteControlServer.Command.PRIORITY_HIGH
                tcp.sendDataRetry(command)
            Case RemoteControlServer.Command.PRIORITY_INDISPENSABLE
                tcp.sendDataUntilReceived(command)
            Case Else
                tcp.sendData(command)
        End Select
    End Sub

    Public Sub sendCommand(ByVal command As Command)
        Dim sendThread As Thread = New Thread(Sub() sendCommandThread(command))
        sendThread.Start()
    End Sub

    Public Sub getWebRequestAsync(ByVal url As String, ByVal callback As Func(Of String, Boolean))
        Dim requestThread As Thread = New Thread(Sub() getWebRequest(url, callback))
        requestThread.Start()
    End Sub

    Private Sub getWebRequest(ByVal url As String, ByVal callback As Func(Of String, Boolean))
        callback(getWebRequest(url))
    End Sub

    Public Function getWebRequest(ByVal url As String) As String
        Dim result = ""
        Try
            Dim request As WebRequest = WebRequest.Create(url)
            request.Timeout = 5000

            Dim response As WebResponse = request.GetResponse()
            Dim dataStream As Stream = response.GetResponseStream()
            Dim reader As New StreamReader(dataStream)
            Dim responseFromServer As String = reader.ReadToEnd()
            result = responseFromServer
            reader.Close()
            response.Close()
        Catch ex As Exception
            Logger.add("Unable to get webrequest: " & url)
        End Try
        Return result
    End Function

    Public Sub loadInBrowser(ByVal url As String)
        Server.gui.loadInBrowser(url)
    End Sub

    Public Function isValidIp(ByVal ip As String) As Boolean
        Try
            System.Net.IPAddress.Parse(ip)
            Return True
        Catch ex As Exception
            Return False
        End Try
    End Function

End Module

File: Source/MacOS/RemoteControlServer/network.cpp
#include "logger.h"
#include "network.h"
#include "remote.h"
#include "server.h"
#include "tcp.h"
#include "udp.h"

#include <QDesktopServices>
#include <QNetworkAccessManager>
#include <QNetworkInterface>
#include <QNetworkRequest>
#include <QtConcurrent>

#include <QDebug>

Network* Network::instance = NULL;

Network* Network::Instance()
{
    if (!instance)
    {
        instance = new Network();
    }
    return instance;
}

Network::Network():
    localHost("127.0.0.1")
{
    Logger::Instance()->add("Initializing network");
    hostIps = new QStringList();
    getHostIps();
    TCP::Instance();
    UDP::Instance();

    commandCount = 0;

    QtConcurrent::run(this, &Network::checkListenersRunning);

    connect(this, SIGNAL(sendData(Command*)), TCP::Instance(), SLOT(sendData(Command*)));
    connect(this, SIGNAL(sendDataRetry(Command*)), TCP::Instance(), SLOT(sendDataRetry(Command*)));
    connect(this, SIGNAL(sendDataUntilReceived(Command*)), TCP::Instance(), SLOT(sendDataUntilReceived(Command*)));
}

bool Network::checkListenersRunning()
{
    if (TCP::Instance()->isListening && UDP::Instance()->isListening)
    {
        return true;
    }
    else
    {
        Logger::Instance()->add("Network initialization failed, not all listeners could be started.");
        Server::Instance()->showErrorDialog("Error", QString("Network initialization failed, not all listeners could be started.\n")
                                            + QString("Make sure that only one instance of the Remote Control Server is running."));
        return false;
    }
}

void Network::getHostIps()
{
    Logger::Instance()->add("Active Wi-Fi network adapters:");
    foreach (const QHostAddress &address, QNetworkInterface::allAddresses())
    {
        if (address.protocol() == QAbstractSocket::IPv4Protocol && address != QHostAddress::LocalHost)
        {
            QString ip = address.toString();
            hostIps->append(ip);
            Logger::Instance()->add(" - " + ip);
        }
    }
}

QString Network::getServerIp()
{
    QString ip = "Unknown";
    for (int i = 0; i < hostIps->length(); ++i)
    {
        if (hostIps->at(i).startsWith("192.168") || hostIps->at(i).startsWith("10.") || hostIps->at(i).startsWith("172."))
        {
            ip = hostIps->at(i);
            break;
        }
    }
    if (Remote::Instance()->lastCommandIsInitialized && hostIps->contains(Remote::Instance()->lastCommand->destination))
    {
        ip = Remote::Instance()->lastCommand->destination;
    }
    return ip;
}

void Network::sendCommand(Command &command)
{
    switch (command.priority)
    {
    case Command::PRIORITY_LOW:
//        UDP::Instance()->sendData(command);
        // send Date via UDP (Not implemented, nor used)
        break;
    case Command::PRIORITY_MEDIUM:
        emit sendData(&command);
        break;
    case Command::PRIORITY_HIGH:
        emit sendDataRetry(&command);
        break;
    case Command::PRIORITY_INDISPENSABLE:
        emit sendDataUntilReceived(&command);
        break;
    default:
        emit sendData(&command);
    }
}

void Network::loadInBrowser(QString &url)
{
    QNetworkAccessManager *manager = new QNetworkAccessManager();
    manager->get(QNetworkRequest(QUrl(url)));
}

bool Network::isValidIp(QString &ip)
{
    QHostAddress ipAddress;
    return (ipAddress.setAddress(ip)) ? true : false;
}
